bucket.1
.TH BUCKET 1
.SH NAME
bucket \- a clipboard manager or registers for the shell
.SH SYNOPSIS
command | \fBbucket\fR [\fIOPTIONS\fR] [\fIBUCKET\fR]
.PP
\fBbucket\fR [\fIOPTIONS\fR] [\fIBUCKET\fR] [\fIDATA\fR]
.PP
The first form reads data from \fISTDIN\fR and writes to the default bucket, or \fIBUCKET\fR if given.
The second form reads \fIDATA\fR and/or \fIBUCKET\fR from arguments, writing to a bucket or printing a bucket's contents.
.SH DESCRIPTION
\fBbucket\fR is like Emacs/Vim-style registers for the shell.
It makes it easy to save clips of text to named "buckets" (or a default, unnamed bucket) and access them later.
.SH OPTIONS
.TP
\fB\-a\fR, \fB\-\-append\fR
Append to bucket
.TP
\fB\-d\fR, \fB\-\-date\fR
Sort by date
.TP
\fB\-e\fR, \fB\-\-edit\fR
Edit bucket
.TP
\fB\-E\fR, \fB\-\-empty\fR
Empty bucket
.TP
\fB\-g\fR \fIPATTERN\fR, \fB\-\-grep\fR \fIPATTERN\fR
Grep in buckets
.TP
\fB\-h\fR, \fB\-\-help\fR
i can Haz cheezburger?
.TP
\fB\-l\fR, \fB\-\-list\fR
List buckets
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Verbose output
.TP
\fB\-V\fR, \fB\-\-VERBOSE\fR
VERY verbose output
.TP
\fB\-x\fR, \fB\-\-expire\fR
eXpire old buckets (default: +14 days)
.TP
\fB\-\-directory\fR \fIDIRECTORY\fR
Bucket storage directory
.SH EXAMPLES
Pour into default bucket by argument:
.PP
.nf
.RS
$ \fBbucket\fR "I wondered why the frisbee was getting bigger, and then it hit me"
.RE
.fi
.PP
Pour from default bucket:
.PP
.nf
.RS
$ \fBbucket\fR
I wondered why the frisbee was getting bigger, and then it hit me
.RE
.fi
.PP
Pipe default bucket into named bucket:
.PP
.nf
.RS
$ \fBbucket\fR | \fBbucket\fR jokes
.RE
.fi
.PP
Pipe into named bucket, verbosely:
.PP
.nf
.RS
$ dmesg | grep error | \fBbucket\fR -v errors
[    7.817768] EXT4-fs (sdc1): re-mounted. Opts: errors=remount-ro,user_xattr
.RE
.fi
.PP
Pour from named bucket:
.PP
.nf
.RS
$ \fBbucket\fR errors
[    7.817768] EXT4-fs (sdc1): re-mounted. Opts: errors=remount-ro,user_xattr
.RE
.fi
.PP
Append to a bucket by argument:
.PP
.nf
.RS
$ \fBbucket\fR -a jokes "A plateau is the highest form of flattery."
$ \fBbucket\fR jokes
I wondered why the frisbee was getting bigger, and then it hit me
A plateau is the highest form of flattery. 
.RE
.fi
.PP
List buckets:
.PP
.nf
.RS
$ \fBbucket\fR -l
bucket  errors  jokes
.RE
.fi
.PP
List buckets verbosely (just the first line of each bucket):
.PP
.nf
.RS
$ \fBbucket\fR -lv
bucket: I wondered why the frisbee was getting bigger, and then it hit me
errors: [    7.817768] EXT4-fs (sdc1): re-mounted. Opts: errors=remount-ro,user_xattr
jokes: I wondered why the frisbee was getting bigger, and then it hit me
.RE
.fi
.PP
List buckets completely:
.PP
.nf
.RS
$ \fBbucket\fR -lV
bucket:
I wondered why the frisbee was getting bigger, and then it hit me
errors: 
[    7.817768] EXT4-fs (sdc1): re-mounted. Opts: errors=remount-ro,user_xattr
jokes:
I wondered why the frisbee was getting bigger, and then it hit me
A plateau is the highest form of flattery. 
.RE
.fi
.PP
Grep buckets:
.PP
.nf
.RS
$ \fBbucket\fR -g frisbee
bucket:I wondered why the frisbee was getting bigger, and then it hit me
jokes:I wondered why the frisbee was getting bigger, and then it hit me
.RE
.fi
.PP
Edit buckets (using \fI$EDITOR\fR):
.PP
.nf
.RS
$ \fBbucket\fR -e
Waiting for Emacs...
$ \fBbucket\fR -e jokes
Waiting for Emacs...
.RE
.fi
.PP
Empty buckets:
.PP
.nf
.RS
$ \fBbucket\fR -E
$ \fBbucket\fR
$ \fBbucket\fR -E jokes
$ \fBbucket\fR jokes
.RE
.fi
.PP
Expire old buckets:
.PP
.nf
.RS
$ sleep 14d  # Default: 2 weeks
$ \fBbucket\fR -x
$ \fBbucket\fR -l
.RE
.fi
.SH CONFIGURATION
.B
Empty/expire commands
.RS
The default command for \fI\-\-empty\fR and \fI\-\-expire\fR is \fItrash-put\fR, part of the \fItrash-cli\fR package,which may not be installed on your system by default.
\fItrash-cli\fR is a great package, so I recommend installing it.  It uses the standard XDG trash bin from the terminal.
.PP
If you prefer, you can easily change the default to \fIrm\fR by changing the \fIdeleteCommand\fR variable near the top of the script.  
.PP
Or you could change it to something like \fImv -t ~/.local/share/Trash/files\fR, although that might cause the trashed files to not show up in GUI trash apps, since they would be missing the \fI.trashinfo\fR files.
.PP
.RE
.B
Storage directory
.RS
The default directory for bucket storage is \fI~/.cache/bucket\fR.
\fI~/.cache\fR is probably excluded by most backup programs (it should be), so this seems like a good place for buckets, since buckets are intended to be temporary storage.
.PP
However, you can set a custom directory using  the \fI\-\-directory\fR option.
Unlike the default directory, it will not be created for you if it doesn't exist.
You might set this option in an alias, that way whenever you use that alias, the buckets would be stored in that directory.
This could be helpful if you wanted to store a certain set of buckets more permanently, in a place where they would typically be included in backups.
.SH SEE ALSO
For more amusing info, see the readme file at \fI/usr/{local,}/share/doc/bucket/README.org\fR
.SH DEVELOPMENT
I can't think of much else to add right now.
But I welcome suggestions and pull requests.
Bug reports may be kept to yourself (just kidding).
.PP
\fBbucket\fR's home is at \fIhttps://github.com/alphapapa/bucket\fR.
.SH LICENSE
I really don't think a license is necessary for something this straightfoward.
But consider it GPL'ed.


bucket.sh
#!/bin/bash

# * bucket.sh

# ** Defaults
bucket=bucket  # Default bucket name
dir=~/.cache/bucket
deleteCommand=trash-put
expireDays=14


# ** Functions
function debug {
    [[ $debug ]] && echo "DEBUG: $@" >&2
}
function die {
    echo "$@" >&2
    exit 1
}
function verbose {
    [[ $verbose ]] && echo "$@" >&2
}
function usage {
    echo "Usage:"
    echo "    command | bucket [OPTIONS] [BUCKET]"
    echo "    bucket [OPTIONS] [BUCKET] [DATA]"
    echo

    echo "The first form reads data from STDIN and writes to the default bucket,
or BUCKET if given.  The second form reads DATA and/or BUCKET from
arguments, writing to a bucket or printing a bucket's contents."

    echo
    echo "Options:"
    echo "    -a, --append                Append to bucket"
    echo "    -d, --date                  Sort by date"
    echo "    -e, --edit                  Edit bucket"
    echo "    -E, --empty                 Empty bucket"
    echo "    -g PATTERN, --grep PATTERN  Grep in buckets"
    echo "    -h, --help                  i can Haz cheezburger?"
    echo "    -l, --list                  List buckets"
    echo "    -v, --verbose               Verbose output"
    echo "    -V, --VERBOSE               VERY verbose output"
    echo "    -x, --expire                eXpire old buckets (default: +$expireDays days)"
    echo "    --directory DIRECTORY       Bucket storage directory"
}


# ** Args
args=$(getopt -o adDeEghlvVx -l "append,date,debug,edit,empty,grep,help,list,verbose,VERBOSE,expire,directory:" -n "bucket" -- "$@")
[[ $? -eq 0 ]] || exit 1

eval set -- "$args"

while true
do
    case "$1" in
        -a|--append)
            append=true ;;
        -d|--date)
            sort=-tr ;;
        -D|--debug)
            debug=true ;;
        -e|--edit)
            edit=true ;;
        -E|--empty)
            empty=true ;;
        -g|--grep)
            shift
            grep="$@"
            break ;;
        -h|--help)
            usage
            exit ;;
        -l|--list)
            list=true ;;
        -v|--verbose)
            verbose=-v ;;
        -V|--VERBOSE)
            verbose=-v
            reallyVerbose=true ;;
        -x|--expire)
            expire=true ;;
        --directory)
            shift
            customDir="$1"
            ;;
        --)
            # Remaining args
            shift
            args=("$@")
            numargs=${#args[@]}
            break ;;
    esac

    shift
done

# *** Check STDIN and bucket name
if ! [[ -t 0 ]]
then
    debug "Data from STDIN"

    stdin=true

    if [[ $numargs -eq 0 ]]
    then
        debug "No args; using default bucket"

    elif [[ $numargs -eq 1 ]]
    then
        debug "One arg; using bucket: $optarg"

        bucket=$args

    else
        debug "Multiple args; using bucket $optarg[1]"
        verbose "Ignoring extra arguments"

        bucket="${args[0]}"
    fi
else
    debug "No data from STDIN"

    if [[ $numargs -eq 0 ]]
    then
        debug "No args"

    elif [[ $numargs -eq 1 ]]
    then
        if [[ -f $dir/$args ]]
        then
            # Bucket exists with name; pour it
            bucket=$args

            debug "Using bucket $bucket; no data remaining in args"
        else
            # No such bucket; use args as data for default bucket
            data=$args

            debug "No bucket named '$args'; using args as data for default bucket"
        fi

    else
        # Multiple args
        bucket="${args[0]}"
        unset args[0]
        data="${args[@]}"

        debug "Using bucket $bucket; using $numargs remaining args as data"
    fi
fi

# **** Sanitize bucket name (since it's passed to eval and trash-put/rm)
bucket=$(echo "$bucket" | sed -r 's/[~.]//g')

# *** Check for conflicting args
if [[ $empty && $expire ]]
then
    conflicting=true
fi
if [[ $conflicting ]]
then
    die "Conflicting operations given."
fi

# ** Check directory
if [[ $customDir ]]
then
    # Custom bucket directory; don't make it
    [[ -d $customDir ]] || die "Directory doesn't exist: $customDir"

    dir=$customDir

else
    # Standard bucket directory; make it if necessary
    if ! [[ -d $dir ]]
    then
        # Make dir
        mkdir -p "$dir" || die "Unable to make bucket directory: $dir"
    fi
fi

# cd just to be extra safe
cd "$dir" || die "Unable to enter bucket directory: $dir"


# ** Main
if [[ $list ]]
then
    # *** List buckets
    readarray -t files <<<"$(ls $sort)"

    if [[ $verbose && ! $reallyVerbose ]]
    then
        for file in "${files[@]}"
        do
            echo "$file: $(head -n1 "$file")"
        done
    elif [[ $reallyVerbose ]]
    then
        for file in "${files[@]}"
        do
            echo "$file:"
            cat "$file"
            echo
        done
    else
        ls
    fi

elif [[ $grep ]]
then
    grep -i "$grep" ./*

elif [[ $edit ]]
then
    # *** Edit bucket
    eval "$EDITOR \"$dir/$bucket\""

elif [[ $empty ]]
then
    # *** Empty bucket
    debug "Emptying bucket..."

    if ! [[ -f $dir/$bucket ]]
    then
        verbose "No such bucket"
    else
        eval "$deleteCommand $verbose \"$dir/$bucket\""
    fi

elif [[ $expire ]]
then
    # *** Expire buckets
    find "$dir" -type f -mtime +$expireDays -exec "$deleteCommand $verbose" '{}' +

else
    if ! [[ $stdin || $data ]]
    then
        # *** Pour bucket
        debug "Pouring bucket..."

        if ! [[ -r $dir/$bucket ]]
        then
            verbose "No such bucket"
        elif ! [[ -s $dir/$bucket ]]
        then
            verbose "Bucket is empty"
        else
            cat "$dir/$bucket"
        fi

    else
        # *** Fill bucket
        debug "Filling bucket..."

        if [[ $stdin ]]
        then
            debug "Catting STDIN into bucket"

            if [[ $append ]]
            then
                cat >>"$dir/$bucket"
            else
                cat >"$dir/$bucket"
            fi

        else
            debug "Echoing data into bucket"

            if [[ $append ]]
            then
                echo "$data" >>"$dir/$bucket"
            else
                echo "$data" >"$dir/$bucket"
            fi

        fi

        # Display bucket if verbose
        [[ $verbose ]] && cat "$dir/$bucket"
    fi
fi
